concent_api/core/payments/report.py

```diff
@@ -0,0 +1,759 @@
+    amount: int = 1,
+) -> None:
+    """ Add error if there are more messages matched with a payment than given amount. """
+    if len(matched_objects) > amount:
+        errors.append(
+            Error(
+                matching_error=MatchingError.MORE_THAN_EXPECTED_MATCHING_MESSAGES_FOR_PAYMENT,
+                message=f'Payment of type {type(payment)} and tx_hash {payment.tx_hash} has {len(matched_objects)} matching messages.',
+            )
+        )
+
+
+def check_error_more_matching_payment_for_message_than_given_amount(
+    errors: List[Error],
+    matched_objects: List,
+    pending_response: PendingResponse,
+    amount: int = 1,
+) -> None:
+    """ Add error if there are more payments matched with a message than given amount. """
+    if len(matched_objects) > amount:
+        errors.append(
+            Error(
+                matching_error=MatchingError.MORE_THAN_ONE_MATCHING_PAYMENT_FOR_MESSAGE,
+                message=f'PendingResponse of type {pending_response.response_type} and pk {pending_response.pk} has {len(matched_objects)} matching payments.',
+            )
+        )
+
+
+def check_error_message_value_is_different_than_payment_value(
+    errors: List[Error],
+    matched_objects: List,
+    force_payment_committed_pending_response: PendingResponse,
+) -> None:
+    """
+    Add error if PaymentInfo.amount_paid related to stored PendingResponse of type ForcePaymentCommitted
+    differs from the settlement_payment.amount value.
+    """
+    payment_info = force_payment_committed_pending_response.payments.order_by('id').last()
+    for settlement_payment in matched_objects:
+        if settlement_payment.amount != payment_info.amount_paid:
+            errors.append(
+                Error(
+                    matching_error=MatchingError.MESSAGE_VALUE_DIFFERS_FROM_PAYMENT_VALUE,
+                    message=f'PaymentInfo with pk {payment_info.pk} amount_paid value {payment_info.amount_paid} '
+                            f'related to stored PendingResponse of type ForcePaymentCommitted '
+                            f'with pk {force_payment_committed_pending_response.pk} '
+                            f'differs from the settlement payment with tx_hash {settlement_payment.tx_hash} '
+                            f'amount value {settlement_payment.amount}.',
+                )
+            )
+
+
+def check_error_message_payment_ts_is_different_than_payment_closure_time(
+    errors: List[Error],
+    matched_objects: List,
+    force_payment_committed_pending_response: PendingResponse,
+) -> None:
+    """
+    Add error if PaymentInfo.payment_ts related stored PendingResponse of type ForcePaymentCommitted
+    differs from closure_time of the settlement payment.
+    """
+    payment_info = force_payment_committed_pending_response.payments.order_by('id').last()
+    for settlement_payment in matched_objects:
+        if settlement_payment.closure_time != parse_datetime_to_timestamp(payment_info.payment_ts):
```

Comment:
I think you should assert that `parse_datetime_to_timestamp(payment_info.payment_ts)` does not have fractional part. In some places you convert timestamps to `int`, in others you don't. If they do have fractional parts, the results will be inconsistent.

Reply:
@cameel, the problem is that `payment.closure_time` is int and all timestamps could be float... we shouldn't assert because timestamps **could** be floats...

Reply:
I'm assuming it's being validated elsewhere (e.g. in golem-messages) If not I think it should be. Or are golem-messages allowed to contain fractional times? The rule we agreed on early in the project for validation is that anything that does not fail whan passed to `int()` is a valid timestamp (see https://github.com/golemfactory/concent/issues/54#issuecomment-358923252). Has GM dropped this rule?